Support for specifying external LoadBalancerIP on openstack

FengyunPan opened the pull request:
1. Support ServiceAnnotationLoadBalancerFloatingNetworkId for LB v1

2. Support for specifying external LoadBalancerIP on openstack
    Add ServiceAnnotationLoadBalancerInternal annotation to distinguish
    between internal LoadBalancerIP and external LoadBalancerIP.


**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
Fix #50851 

**Release note**:
```release-note
NONE
```

Commit:
Support ServiceAnnotationLoadBalancerFloatingNetworkId for LB v1

CaoShuFeng commented:
/ok-to-test

FengyunPan commented:
/test pull-kubernetes-kubemark-e2e-gce

dims commented:
/assign @anguslees

dims commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -68,6 +68,11 @@ const (
+	// ServiceAnnotationLoadBalancerInternal is the annotation used on the service
+	// to indicate that we want an internal loadbalancer service.
+	// If the value of ServiceAnnotationLoadBalancerInternal is false, it indicates that we want an external loadbalancer service. Default to true.
+	ServiceAnnotationLoadBalancerInternal = "service.beta.kubernetes.io/openstack-load-balancer-internal"
```
Any reason to pick the new format? (compare with line 70)

FengyunPan replied:
ServiceAnnotationLoadBalancerFloatingNetworkId is the annotation used on floating-network-id, the same as 'floating-network-id' which from cloud provider config file.
ServiceAnnotationLoadBalancerInternal is the annotation used to distinguish between internal LoadBalancerIP and external LoadBalancerIP.
If it is true, the service.Spec.LoadBalancerIP is a vip of neutron LB.
If it is false,  the service.Spec.LoadBalancerIP is a floatingip of neutron LB's vip.

Currently openstack cloud provider just support for specifying internal LoadBalancerIP(vip). I think that is unreasonable, actually user just use a floating ip to access a LoadBalancer service. So it is necessary to spport for specifying external LoadBalancerIP(floatingip).

dims replied:
Sorry i think i should have been more specific, looking at line 70, can we use `loadbalancer.openstack.org/openstack-internal-load-balancer` instead of `service.beta.kubernetes.io/openstack-load-balancer-internal`

FengyunPan replied:
All right, I update it soon.

dims commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -626,6 +631,26 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+	switch serviceAnnotationLoadBalancerInternal {
+	case "true":
+		internalAnnotation = true
+		glog.V(4).Infof("Ensure an internal loadbalancer service with ServiceAnnotationLoadBalancerInternal annotation: %v",
```
avoid variable name in the user visible log - use the actual annotation they should use

FengyunPan replied:
Right, done.

dims commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -626,6 +631,26 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+	case "false":
+		internalAnnotation = false
+		if len(floatingPool) != 0 {
+			glog.V(4).Infof("Ensure an external loadbalancer service with ServiceAnnotationLoadBalancerInternal annotation: %v",
```
same as above

FengyunPan replied:
Done.

dims commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -626,6 +631,26 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+			glog.V(4).Infof("Ensure an external loadbalancer service with ServiceAnnotationLoadBalancerInternal annotation: %v",
+				internalAnnotation)
+		} else {
+			return nil, fmt.Errorf("floating-network-id or ServiceAnnotationLoadBalancerFloatingNetworkId should be specified when ServiceAnnotationLoadBalancerInternal is false")
```
same as above

FengyunPan replied:
Done.

dims commented on pkg/cloudprovider/providers/openstack/openstack_loadbalancer.go:
```diff
@@ -626,6 +631,26 @@ func (lbaas *LbaasV2) EnsureLoadBalancer(clusterName string, apiService *v1.Serv
+			return nil, fmt.Errorf("floating-network-id or ServiceAnnotationLoadBalancerFloatingNetworkId should be specified when ServiceAnnotationLoadBalancerInternal is false")
+		}
+	default:
+		return nil, fmt.Errorf("unknow ServiceAnnotationLoadBalancerInternal annotation: %v, specify \"true\" or \"false\".",
```
same as above

FengyunPan replied:
Done.

Commit:
Support for specifying external LoadBalancerIP on openstack

Fix #50851
Add ServiceAnnotationLoadBalancerInternal annotation to distinguish
between internal LoadBalancerIP and external LoadBalancerIP.

dims commented:
/retest

dims commented:
/retest

FengyunPan commented:
/test pull-kubernetes-e2e-kops-aws

xsgordon commented:
/sig openstack

dims commented:
/approve
/lgtm
